### ISSUE TITLE
[🐛 Bug]: Error while loading the running and queued Sessions from the Grid 

### ISSUE THREAD

**Issue by lawajs** [SeleniumHQ/selenium#9988]
### What happened?

I am getting an error when running webdriver.io based regression test suite. In this case, we have selenium grid and the browser nodes' instances set up in the Kubernetes cluster, there are 3 browser types consisting of chrome, edge and firefox. I set the test up to run against 2 instances of chrome and each time the test is fired it spins up 2 instances of chrome and runs about 4 tests before it crashes with the error message as: 
"There has been an error while loading the running and queued Sessions from the Grid.
Unexpected token < in JSON at position 0"

Before this, I was getting spurious session queues with the session not cleaning up after the test might have stopped and eventual crashes. I managed this by including the following in my hook file:

afterSession: async function() {
        // workaround to make sure the chromedriver shuts down
        await browser.end().pause(1000);
    },

    after: async function() {
        // workaround to make sure the chromedriver shuts down
        await browser.pause(1000);
    },
Then, I started getting the crashes as described above

### How can we reproduce the issue?

```shell
This can be reproduced by setting up the selenium grid with the browsers' nodes in the Kubernetes cluster (in my case, I was testing against a remote selenium grid setup) and attempt to run a suite of webdriver.io cucumber scenarios based tests (12 or more tests) against it.
```

### Relevant log output

```shell
There has been an error while loading the running and queued Sessions from the Grid.
Unexpected token < in JSON at position 0
```

### Operating System

macOS BigSur

### Selenium version

Selenium grid: 4.0.0-20211013, WebDriver.io, Nodejs vs 14.17.1

### What are the browser(s) and version(s) where you see this issue?

Chrome 94.0.4606.81

### What are the browser driver(s) and version(s) where you see this issue?

ChromeDriver 94.0.4606.81

### Are you using Selenium Grid?

Selenium grid: 4.0.0

**Comment by amccague** [comment]
To be specific with some of the errors (I am helping @lawajs with his issue) - it appears that the /graphql endpoint on the hub has totally crashed, it always times out, even when only querying version - so the error he is seeing is a symptom of the UI parsing this timeout.

`/wd/hub/status` and `/status` is still fine and reporting as normal.

**Comment by AutomatedTester** [comment]
I think this is fixed in 6bbafcdde8ac904b9bf9f88e8c9e9423c0d5d99d. @pujagani can you confirm?

**Comment by pujagani** [comment]
Thank you for sharing the details. @AutomatedTester If the error is a due to bad event-bus message then yes, `6bbafcd` fixes it. However, based on what I understand so far, the error message shared is related to the Grid UI unable to parse the message related to the session. The error message shared is defined in `https://github.com/SeleniumHQ/selenium/blob/64447d4b03f6986337d1ca8d8b6476653570bcc1/javascript/grid-ui/src/screens/Sessions/Sessions.tsx#L102 `. However, I am unable to understand what exactly is failing and I have not seen something like this locally as well. 
@lawajs @amccague I would really appreciate it if you provide a complete reproduction script, please. It will help us identify if the error is due to a bad message or do we need to fix our error handling or something else is causing things to hang.  Also, if you can share the logs from the Grid components that would be a great.

**Comment by pujagani** [comment]
Does https://github.com/SeleniumHQ/selenium/issues/9959 this sound familiar to you @lawajs @amccague? If there is something, not Selenium-related, pinging the pods then it could be the event bus just hanging and not able to process future events, which can also prevent session clean up. If that is the case, then it is solved and will be available in the next release.

**Comment by amccague** [comment]
@pujagani yes the behaviour does look similar. This is the only WARN/ERROR I can locate that coincides with what we're seeing (prettified slightly):
```
[SeleniumSpanExporter$1.lambda$export$0] - {
  "traceId": "d1672860cc6c5328eceb8e4ae7c20eb0",
  "eventTime": 1635264409429872000,
  "eventName": "exception",
  "attributes": {
    "exception.message": "Unable to execute request for an existing session: Unable to find session with ID: 69c3211f7ce087437d02eaf775cf9663
Build info: version: '4.0.0', revision: '3a21814679'
System info: host: 'selenium-hub-78f5fb6678-tksfs', ip: '10.16.58.238', os.name: 'Linux', os.arch: 'amd64', os.version: '5.4.141-67.229.amzn2.x86_64', java.version: '11.0.11'
Driver info: driver.version: unknown",
    "exception.stacktrace": "org.openqa.selenium.NoSuchSessionException: Unable to find session with ID: 69c3211f7ce087437d02eaf775cf9663
Build info: version: '4.0.0', revision: '3a21814679'
System info: host: 'selenium-hub-78f5fb6678-tksfs', ip: '10.16.58.238', os.name: 'Linux', os.arch: 'amd64', os.version: '5.4.141-67.229.amzn2.x86_64', java.version: '11.0.11'
Driver info: driver.version: unknown
   at org.openqa.selenium.grid.sessionmap.local.LocalSessionMap.get(LocalSessionMap.java:129)
   at org.openqa.selenium.grid.router.HandleSession.lambda$loadSessionId$3(HandleSession.java:147)
   at io.opentelemetry.context.Context.lambda$wrap$2(Context.java:219)
   at org.openqa.selenium.grid.router.HandleSession.execute(HandleSession.java:110)
   at org.openqa.selenium.remote.http.Route$PredicatedRoute.handle(Route.java:373)
   at org.openqa.selenium.remote.http.Route.execute(Route.java:68)
   at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:336)
   at org.openqa.selenium.remote.http.Route.execute(Route.java:68)
   at org.openqa.selenium.grid.router.Router.execute(Router.java:91)
   at org.openqa.selenium.grid.web.EnsureSpecCompliantResponseHeaders.lambda$apply$0(EnsureSpecCompliantResponseHeaders.java:34)
   at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:64)
   at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:336)
   at org.openqa.selenium.remote.http.Route.execute(Route.java:68)
   at org.openqa.selenium.remote.http.Route$NestedRoute.handle(Route.java:270)
   at org.openqa.selenium.remote.http.Route.execute(Route.java:68)
   at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:336)
   at org.openqa.selenium.remote.http.Route.execute(Route.java:68)
   at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:336)
   at org.openqa.selenium.remote.http.Route.execute(Route.java:68)
   at org.openqa.selenium.remote.AddWebDriverSpecHeaders.lambda$apply$0(AddWebDriverSpecHeaders.java:35)
   at org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)
   at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:64)
   at org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)
   at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:64)
   at org.openqa.selenium.netty.server.SeleniumHandler.lambda$channelRead0$0(SeleniumHandler.java:44)
   at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
   at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
   at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
   at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
   at java.base/java.lang.Thread.run(Thread.java:829)
",
    "exception.type": "org.openqa.selenium.NoSuchSessionException",
    "http.client_ip": "10.120.253.3",
    "http.flavor": 1,
    "http.handler_class": "org.openqa.selenium.grid.router.HandleSession",
    "http.host": "[redacted]",
    "http.method": "POST",
    "http.request_content_length": "8299",
    "http.scheme": "HTTP",
    "http.target": "/session/69c3211f7ce087437d02eaf775cf9663/execute/sync",
    "http.user_agent": "webdriver/7.14.1",
    "session.id": "69c3211f7ce087437d02eaf775cf9663"
  }
}
```

When will the next release be cut?

**Comment by pujagani** [comment]
Thank you for the response and thank you for sharing the logs. Based on the logs, it seems like the client script is trying to access an already deleted session (i.e driver.quit is called and then the same driver is used to send new commands). I would recommend just double checking if the client-side script to avoid such an error in the future. 
I have a hunch the core problem could be event bus related. Afaik, we are close to the next release. The next release should be in a week or two. We are trying to ensure some critical issues are fixed before that.

**Comment by lawajs** [comment]
@pujagani Many thanks for your response. Interestingly, I didn't use driver quit in my script neither did I used session delete. The browser.end command that I used was used in the afterSession and the after section. I thought using those will help to clean things up however, even when they are not being used, I was still getting the error. Apart from that, I guess when using Webdriver.io, and wdio test runner one can access the webdriver instance through the global browser object. The session is initialized by the test runner so one doesn't need to call init command. The same goes for ending the session. This is also done by the test runner process > https://webdriver.io/docs/browserobject/

**Comment by pujagani** [comment]
Thank you for providing a detailed response. I get the context now. This can also happen when a Node is marked down by the Hub since all the existing sessions are deleted at that time. If a request is made then for a particular session since the session does not exist, then it throws a `NoSuchSessionException ` and logs a message `"exception.message": "Unable to execute request for an existing session: Unable to find session with ID: <session-id>`. Have you observed a similar pattern everytime this log statement appears? If so, then I am afraid something like this https://github.com/SeleniumHQ/selenium/issues/9594 is happening again.

**Comment by diemol** [comment]
@lawajs @amccague 
I am quite certain this is related to the EventBus, and then it should have been fixed with the commit mentioned above.

Just to get more information, can you share the complete logs of the Nodes?

Also, I just made a prerelease where you can test this fix
https://github.com/SeleniumHQ/docker-selenium/releases/tag/4.1.0-prerelease-20211105

**Comment by lawajs** [comment]
@diemol Many thanks, I will try the prerelease and get back to you.

**Comment by kitoymonster** [comment]
Hi, we are also experiencing this issue in our remote server. The selenium server dies when running 3 or more threads. I'm not sure if this is related or even the same issue, but I tried to reproduce this locally by running a test suite with 8 threads.

For some reason, the session got deleted. No error prior to that. And it seems the test related to that session is still running as shown in the 3rd entry. Not sure if this belongs here or if this should be filed as a different issue.

```
21:58:15.851 INFO [LocalDistributor.newSession] - Session created by the distributor. Id: 0d7d3589ef8f8f38465da0c3c34f4c0d, Caps: Capabilities {acceptInsecureCerts: true, browserName: chrome, browserVersion: 95.0.4638.69, chrome: {chromedriverVersion: 95.0.4638.69 (6a1600ed572fe..., userDataDir: /var/folders/h2/68fwny6s7gx...}, goog:chromeOptions: {debuggerAddress: localhost:53195}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: mac os x, proxy: Proxy(), se:cdp: ws://192.168.254.100:4444/s..., se:cdpVersion: 95.0.4638.69, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true}

22:03:58.976 INFO [LocalSessionMap.lambda$new$0] - Deleted session from local session map, Id: 0d7d3589ef8f8f38465da0c3c34f4c0d

22:04:35.390 WARN [SeleniumSpanExporter$1.lambda$export$0] - {"traceId": "c1f6562d7eb958e3ad9fff7d26b045d0","eventTime": 1636466675389629758,"eventName": "exception","attributes": {"exception.message": "Unable to execute request for an existing session: Unable to find session with ID: 0d7d3589ef8f8f38465da0c3c34f4c0d
Build info: version: '4.0.0', revision: '3a21814679'
System info: host: 'Keith-MacBook-Pro.local', ip: 'fe80:0:0:0:c35:c65e:feb3:70c6%en0', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.16', java.version: '15'
Driver info: driver.version: unknown","exception.stacktrace": "org.openqa.selenium.NoSuchSessionException: Unable to find session with ID: 0d7d3589ef8f8f38465da0c3c34f4c0d
Build info: version: '4.0.0', revision: '3a21814679'
System info: host: 'Keith-MacBook-Pro.local', ip: 'fe80:0:0:0:c35:c65e:feb3:70c6%en0', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.16', java.version: '15'
Driver info: driver.version: unknown
	at org.openqa.selenium.grid.sessionmap.local.LocalSessionMap.get(LocalSessionMap.java:129)
  at org.openqa.selenium.grid.router.HandleSession.lambda$loadSessionId$3(HandleSession.java:147)
  at io.opentelemetry.context.Context.lambda$wrap$2(Context.java:219)
  at org.openqa.selenium.grid.router.HandleSession.execute(HandleSession.java:110)
  at org.openqa.selenium.remote.http.Route$PredicatedRoute.handle(Route.java:373)
  at org.openqa.selenium.remote.http.Route.execute(Route.java:68)
  at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:336)
  at org.openqa.selenium.remote.http.Route.execute(Route.java:68)
  at org.openqa.selenium.grid.router.Router.execute(Router.java:91)
  at org.openqa.selenium.grid.web.CheckOriginHeader.lambda$apply$0(CheckOriginHeader.java:66)
  at org.openqa.selenium.grid.web.CheckContentTypeHeader.lambda$apply$0(CheckContentTypeHeader.java:70)
  at org.openqa.selenium.grid.web.EnsureSpecCompliantResponseHeaders.lambda$apply$0(EnsureSpecCompliantResponseHeaders.java:34)
  at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:64)
  at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:336)
  at org.openqa.selenium.remote.http.Route.execute(Route.java:68)
  at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:336)
  at org.openqa.selenium.remote.http.Route.execute(Route.java:68)
  at org.openqa.selenium.remote.AddWebDriverSpecHeaders.lambda$apply$0(AddWebDriverSpecHeaders.java:35)
  at org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)
  at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:64)
  at org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)
  at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:64)
  at org.openqa.selenium.netty.server.SeleniumHandler.lambda$channelRead0$0(SeleniumHandler.java:44)
  at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
  at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
  at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
  at java.base/java.lang.Thread.run(Thread.java:832)
  ","exception.type": "org.openqa.selenium.NoSuchSessionException","http.flavor": 1,"http.handler_class": "org.openqa.selenium.grid.router.HandleSession","http.host": "localhost:4444","http.method": "POST","http.request_content_length": "56","http.scheme": "HTTP","http.target": "/session/0d7d3589ef8f8f38465da0c3c34f4c0d/element","http.user_agent": "webdriver/7.16.3","session.id": "0d7d3589ef8f8f38465da0c3c34f4c0d"}}``` 

We also tried 4.0.0 alpha 6, 4.0.0 rc1, and even https://github.com/SeleniumHQ/docker-selenium/releases/tag/4.1.0-prerelease-20211105 but the issue is still reproducible.

**Comment by lawajs** [comment]
@diemol Many thanks again. 
The browsers no longer crashes and running of test against a single instance of a browser is fine with all tests running successfully without any issue however, when running against multiple instances of a browser e.g. Chrome, some of the tests that passed against a single instance are failing. The same issue is experienced when performing multi-browser testing. In this case, no more hanging queues or crashing nodes but, I am still wondering why testing against a single instance of a browser runs all the test successfully and multiple instances of the same browser is failing. I am using https://github.com/SeleniumHQ/docker-selenium/releases/tag/4.1.0-prerelease-20211105

**Comment by Ajay2613** [comment]
@diemol : I am still facing the same issue in 4.1.0 jar i.e. hub is going down after some time and throwing "[DefaultChannelPipeline.onUnhandledInboundException] - An exceptionCaught() event was fired, and it reached at the tail of the pipeline. We are able to execute the executions on nodes, However after certain period of time lets say 12 hours it gives the above exception then i have to again up the hub and node. 
![Exception](https://user-images.githubusercontent.com/26387758/144195845-7fb2887d-43e0-4662-86c5-9e863f9e0845.jpg)

**Comment by diemol** [comment]
@lawajs @ajayk could you please share a test script and detailed steps to reproduce the issue?

**Comment by diemol** [comment]
I will close this since we got no more information or related reports. Please open a new issue if you are bumping into something similar.
